src/reconcile/rec_visibility.c

```diff
@@ -758,3 +758,253 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
+         * prepared transaction IDs are globally visible, need to check the update state as well.
+         */
+        if (F_ISSET(r, WT_REC_VISIBLE_ALL) ? WT_TXNID_LE(r->last_running, txnid) :
+                                             !__txn_visible_id(session, txnid)) {
+            r->update_uncommitted = list_uncommitted = true;
+            continue;
+        }
+        if (upd->prepare_state == WT_PREPARE_LOCKED ||
+          upd->prepare_state == WT_PREPARE_INPROGRESS) {
+            list_prepared = true;
+            if (upd->start_ts > max_ts)
+                max_ts = upd->start_ts;
+            continue;
+        }
+
+        /* Track the first update with non-zero timestamp. */
+        if (upd->durable_ts > max_ts)
+            max_ts = upd->durable_ts;
+
+        /*
+         * Always select the newest committed update to write to disk
+         */
+        if (upd_select->upd == NULL)
+            upd_select->upd = upd;
+
+        if (!__rec_update_durable(session, r, upd)) {
+            if (F_ISSET(r, WT_REC_EVICT))
+                ++r->updates_unstable;
+
+            /*
+             * Rare case: when applications run at low isolation levels, update/restore eviction may
+             * see a stable update followed by an uncommitted update. Give up in that case: we need
+             * to discard updates from the stable update and older for correctness and we can't
+             * discard an uncommitted update.
+             */
+            if (F_ISSET(r, WT_REC_UPDATE_RESTORE) && upd_select->upd != NULL &&
+              (list_prepared || list_uncommitted))
+                return (__wt_set_return(session, EBUSY));
+
+            continue;
+        }
+
+        if (!F_ISSET(r, WT_REC_EVICT))
+            break;
+    }
+
+    /* Keep track of the selected update. */
+    upd = upd_select->upd;
+
+    /* Reconciliation should never see an aborted or reserved update. */
+    WT_ASSERT(
+      session, upd == NULL || (upd->txnid != WT_TXN_ABORTED && upd->type != WT_UPDATE_RESERVE));
+
+    /*
+     * The checkpoint transaction is special. Make sure we never write metadata updates from a
+     * checkpoint in a concurrent session.
+     */
+    WT_ASSERT(session, !WT_IS_METADATA(session->dhandle) || upd == NULL ||
+        upd->txnid == WT_TXN_NONE || upd->txnid != S2C(session)->txn_global.checkpoint_state.id ||
+        WT_SESSION_IS_CHECKPOINT(session));
+
+    /* If all of the updates were aborted, quit. */
+    if (upd == NULL) {
+        return (0);
+    }
+
+    if (upd != NULL && upd->durable_ts > r->max_ondisk_ts)
+        r->max_ondisk_ts = upd->durable_ts;
+
+    /*
+     * TIMESTAMP-FIXME The start timestamp is determined by the commit timestamp when the key is
+     * first inserted (or last updated). The end timestamp is set when a key/value pair becomes
+     * invalid, either because of a remove or a modify/update operation on the same key.
+     */
+    if (upd != NULL) {
```

Comment:
Probably better to move this part outside this function. I think it is enough for this function just to return the selected update. Thoughts? The caller of the function should figure out the start and stop time pairs.

Reply:
I'd prefer it if you hold off from refactoring this part. My change to write to the on-disk validity window touches this part of `__wt_rec_upd_select` so any further changes here are going to make it difficult for me.

Reply:
Understood. I'll keep it as it is for now.